packages/cozy-client/src/CozyClient.js

```diff
@@ -306,38 +309,73 @@ export default class CozyClient {
     return withIncluded
   }
 
-  async fetchRelationships(response, schemaRelsByName) {
+  /**
+   * Fetch relationships for a response (can be several docs).
+   * Fills the `relationships` attribute of each documents.
+   *
+   * Can potentially result in several fetch requests.
+   * Queries are optimized before being sent.
+   */
+  async fetchRelationships(response, relationshipsByName) {
     const isSingleDoc = !Array.isArray(response.data)
     if (!isSingleDoc && response.data.length === 0) {
       return response
     }
     const responseDocs = isSingleDoc ? [response.data] : response.data
 
-    const schemaRels = Object.values(schemaRelsByName)
-    const definitionOrDocuments = flatten(
-      responseDocs.map(doc =>
-        this.prepareQueryDefsFromRelationships(doc, schemaRels)
-      )
-    ).filter(Boolean)
+    // Triplets [docId, relationshipName, relationshipQuery]
+    const docIdRelQueries = flatten(
+      responseDocs.map(doc => {
+        return map(relationshipsByName, (relationship, relName) => {
+          const query = relationship.type.query(doc, this, relationship)
+          return [doc._id, relName, query]
+        })
+      })
+    )
+
+    // Used to reattach responses into the relationships attribute of
+    // each document
+    const queryDefToDocIdAndRel = new Map(
+      docIdRelQueries.map(([docId, relName, queryDef]) => ([queryDef, [docId, relName]]))
+    )
 
-    // Relationships can yield documents, ready for use, or definitions needing
-    // to be executed
+    // Relationships can yield "queries" that are already resolved documents.
+    // These do not need to go through the usual link request mechanism.
     const { documents = [], definitions = [] } = groupBy(
-      definitionOrDocuments,
+      docIdRelQueries.map(triplet => triplet[2]),
       d => (d instanceof QueryDefinition ? 'definitions' : 'documents')
     )
```

Comment:
Are we mapping three times on the "same" content? Is there any space for optimization here?

Reply:
Yes. https://github.com/cozy/cozy-client/pull/232/commits/5ec2e73494970db53892c7c7ed8f9d989239ca7c